libs/ktem/ktem/reasoning/simple.py

```diff
@@ -680,28 +738,90 @@ def stream(  # type: ignore
+            sub_question_answer_str=sub_question_answer_str,
             conv_id=conv_id,
             **kwargs,
         )
 
+        if self.use_rewrite:
+            yield Document(
+                channel="info",
+                content=(
+                    "<b>Rewritten pipeline</b><br>"
+                    f"{rewrite_pipeline.__class__.__name__}<br>"
+                    f"<b>Rewritten question</b><br>{message}<br>"
+                ),
+            )
+
+        yield Document(channel="info", content="<b>Evidence</b><br>")
         # show the evidence
         with_citation, without_citation = self.prepare_citations(answer, docs)
         if not with_citation and not without_citation:
-            yield Document(channel="info", content="No evidence found.\n")
+            yield Document(channel="info", content="No evidence found.<br>")
         else:
             yield Document(channel="info", content=None)
             for _ in with_citation:
                 yield _
             if without_citation:
                 yield Document(
                     channel="info",
-                    content="Retrieved segments without matching evidence:\n",
+                    content="Retrieved segments without matching evidence:<br>",
                 )
                 for _ in without_citation:
                     yield _
 
         return answer
 
+    def answer_sub_questions(
+        self, messages: list, conv_id: str, history: list, **kwargs
+    ):
+        output_str = ""
+        for idx, message in enumerate(messages):
+            yield Document(
+                channel="info",
+                content=f"<b>Sub-question {idx + 1}</b><br>{message}<br>",
+            )
+            # should populate the context
+            docs, infos = self.retrieve(message, history)
+            for _ in infos:
+                yield _
+
+            evidence_mode, evidence = self.evidence_pipeline(docs).content
+            answer = self.answering_pipeline.invoke(
+                question=message,
+                history=history,
+                evidence=evidence,
+                evidence_mode=evidence_mode,
+                conv_id=conv_id,
+                **kwargs,
+            )
+
+            yield Document(channel="info", content="<b>Evidence</b><br>")
+            with_citation, without_citation = self.prepare_citations(answer, docs)
+            if not with_citation and not without_citation:
+                yield Document(channel="info", content="No evidence found.<br>")
+            else:
+                yield Document(channel="info", content=None)
+                for _ in with_citation:
+                    yield _
+                if without_citation:
+                    yield Document(
+                        channel="info",
+                        content="Retrieved segments without matching evidence:<br>",
+                    )
+                    for _ in without_citation:
+                        yield _
```

Comment:
This logic seems to overlap a lot with `ainvoke`. Please refractor the code to reduce unnecessary repetition.